cli_tools/common/utils/logging/logger.go

```diff
@@ -15,50 +15,205 @@
 package logging
 
 import (
-	"fmt"
 	"log"
+	"os"
+	"strings"
+	"sync"
 	"time"
+
+	"github.com/GoogleCloudPlatform/compute-image-tools/proto/go/pb"
+	"github.com/golang/protobuf/proto"
 )
 
-// LogEntry encapsulates a single log entry.
-type LogEntry struct {
-	LocalTimestamp time.Time `json:"localTimestamp"`
-	Message        string    `json:"message"`
+// LogWriter is a logger for interactive tools. It supports
+// string messages and structured metrics.
+//
+// Structured metrics are accumulated over the lifespan of the logger.
+//
+// To rebuild the mock, run `go generate ./...`
+//go:generate go run github.com/golang/mock/mockgen -package mocks -source $GOFILE -destination ../../../mocks/mock_logger.go
+type LogWriter interface {
+	// WriteUser messages appear in the following places:
+	//  1. Web UI and gcloud.
+	//  2. Standard output of the CLI tool.
+	//  3. Backend trace logs (all debug and user logs are combined to a single trace log).
+	WriteUser(message string)
+	// WriteDebug messages appear in the following places:
+	//  1. Standard output of the CLI tool.
+	//  2. Backend trace logs (all debug and user logs are combined to a single trace log).
+	WriteDebug(message string)
+	// WriteTrace messages are saved to the logging backend (OutputInfo.serial_outputs).
+	WriteTrace(message string)
+	// WriteMetric merges all non-default fields into a single OutputInfo instance.
+	WriteMetric(metric *pb.OutputInfo)
+}
+
+// RedirectGlobalLogsToUser redirects the standard library's static logger
+// to logWriter.User.
+func RedirectGlobalLogsToUser(logWriter LogWriter) {
+	log.SetPrefix("")
+	log.SetFlags(0)
+	log.SetOutput(redirectShim{logWriter})
+}
+
+// redirectShim forwards Go's standard logger to LogWriter.User.
+type redirectShim struct {
+	writer LogWriter
+}
+
+func (l redirectShim) Write(p []byte) (n int, err error) {
+	l.writer.WriteUser(string(p))
+	return len(p), nil
+}
+
+// LogReader exposes pb.OutputInfo to a consumer.
+type LogReader interface {
+	ReadOutputInfo() *pb.OutputInfo
```

Comment:
I don't see this called anywhere. What is the plan for this func?

Reply:
The only log consumer is the logging backend -- the `LogReader` interface will replace current usage of `Loggable`:

https://github.com/GoogleCloudPlatform/compute-image-tools/blob/e16e8e02ae7fa1c3e2940987368500b7d821ea61/cli_tools/common/utils/logging/service/logger.go#L184